File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/Gimbal/GimbalCapability/DJIGimbal+CapabilityCheck.m
//
//  DJIGimbal+CapabilityCheck.m
//  DJISdkDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//

#import "DJIGimbal+CapabilityCheck.h"
#import <DJISDK/DJISDK.h>

@implementation DJIGimbal (CapabilityCheck)

-(DJIParamCapability *) getCapabilityWithKey:(NSString *)key {
    if (self.capabilities && self.capabilities[key]) {
        DJIParamCapability *capability = (DJIParamCapability *)self.capabilities[key];
        return capability;
    }
    return nil;
}

-(BOOL) isFeatureSupported:(NSString *)key {
    DJIParamCapability *capability = [self getCapabilityWithKey:key];
    if (capability) {
        return capability.isSupported;
    }
    return NO;
}

-(NSNumber *) getParamMin:(NSString *)key {
    if ([self isFeatureSupported:key]) {
        DJIParamCapability *capability = [self getCapabilityWithKey:key];
        if ([capability isKindOfClass:[DJIParamCapabilityMinMax class]]) {
            DJIParamCapabilityMinMax *capabilityMinMax = (DJIParamCapabilityMinMax *)capability;
            return capabilityMinMax.min;
        }
    }
    return nil;
}

-(NSNumber *) getParamMax:(NSString *)key {
    if ([self isFeatureSupported:key]) {
        DJIParamCapability *capability = [self getCapabilityWithKey:key];
        if ([capability isKindOfClass:[DJIParamCapabilityMinMax class]]) {
            DJIParamCapabilityMinMax *capabilityMinMax = (DJIParamCapabilityMinMax *)capability;
            return capabilityMinMax.max;
        }
    }
    return nil;
}

@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/Gimbal/GimbalCapability/DJIGimbal+CapabilityCheck.swift
//
//  DJIGimbal+CapabilityCheck.swift
//  DJISDKSwiftDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//

import DJISDK
extension DJIGimbal {
    func getCapabilityWithKey(_ key: String) -> DJIParamCapability? {
        if (self.gimbalCapability[key] != nil) {
            let capability: DJIParamCapability = (self.gimbalCapability[key] as! DJIParamCapability)
            return capability
        }
        return nil
    }
    
    func isFeatureSupported(_ key: String) -> Bool {
        let capability: DJIParamCapability = self.getCapabilityWithKey(key)!
        return capability.isSupported
    }
    
    func getParamMin(_ key: String) -> Int? {
        if self.isFeatureSupported(key) {
            let capability: DJIParamCapability = self.getCapabilityWithKey(key)!
            if (capability is DJIParamCapabilityMinMax) {
                let capabilityMinMax: DJIParamCapabilityMinMax = (capability as! DJIParamCapabilityMinMax)
                return capabilityMinMax.min.intValue
            }
        }
        return nil
    }
    
    func getParamMax(_ key: String) -> Int? {
        if self.isFeatureSupported(key) {
            let capability: DJIParamCapability = self.getCapabilityWithKey(key)!
            if (capability is DJIParamCapabilityMinMax) {
                let capabilityMinMax: DJIParamCapabilityMinMax = (capability as! DJIParamCapabilityMinMax)
                return capabilityMinMax.max.intValue
            }
        }
        return nil
    }
}
